atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java

```diff
@@ -76,13 +77,37 @@
      * the batch hint in each RangeRequest. If this isn't done then this method may do more work
      * than you need and will be slower than it needs to be.  If the batchHint isn't specified it
      * will default to 1 for the first page in each range.
+     *
+     * @deprecated Should use either {@link #getUnfetchedRanges(TableReference, Iterable)} or
+     * {@link #getRangesWithFirstPages(TableReference, Iterable, int, ExecutorService)} to ensure you
+     * are using an appropriate level of concurrency.
      */
     @Idempotent
+    @Deprecated
     Iterable<BatchingVisitable<RowResult<byte[]>>> getRanges(
             TableReference tableRef,
             Iterable<RangeRequest> rangeRequests);
 
     /**
+     * Returns visitibles that scan the provided ranges. This does no pre-fetching so visiting the resulting
+     * visitibles will incur database reads on first access.
+     */
+    @Idempotent
+    Iterable<BatchingVisitable<RowResult<byte[]>>> getUnfetchedRanges(
+            final TableReference tableRef, Iterable<RangeRequest> rangeRequests);
+
+    /**
+     * Same as {@link #getUnfetchedRanges(TableReference, Iterable)} but pre-fetches the first page of each visitible.
+     * @param prefetchConcurrency the maximum number of ranges that are pre-fetched concurrently
+     */
+    @Idempotent
+    Iterable<BatchingVisitable<RowResult<byte[]>>> getRangesWithFirstPages(
+            final TableReference tableRef,
+            Iterable<RangeRequest> rangeRequests,
+            int prefetchConcurrency,
+            ExecutorService executorService);
```

Comment:
Yeah main thing I kept going back and forth on is how to do that and also allow for subsequent parallel iteration over the visitables (e.g. get the ranges with prefetched first pages and then concurrently fetch the rest of the pages). Would it make sense to add additional methods in `Transaction` for concurrently operating on the results so we can use the same executor?